### PR TITLE
Add integration tests for Rust BuildMethod

### PR DESCRIPTION
**Pull request by hawflau**
#### Which issue(s) does this change fix?

#4673 

#### Why is this change necessary?
Adding integration tests for Rust build method

#### How does it address the issue?
Add integration tests for two use cases:
- single funtion
- multi binaries

#### What side effects does this change have?
No

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: appveyor-ubuntu.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ environment:
   NODE_VERSION: "14.17.6"
   AWS_S3: 'AWS_S3_TESTING'
   AWS_ECR: 'AWS_ECR_TESTING'
+  CARGO_LAMBDA_VERSION: "v0.17.1"
```
Comment (jfuss) [inline appveyor-ubuntu.yml:26]: Should we run the cargo related tests in GHA instead?

File: tests/integration/buildcmd/test_build_cmd.py [hunk 2]
```diff
@@ -1122,6 +1123,65 @@ def test_go_must_fail_with_unknown_architecture(self, runtime, code_uri, archite
         self.assertEqual(process_execute.process.returncode, 1)
 
 
+@skipIf(
+    ((IS_WINDOWS and RUNNING_ON_CI) and not CI_OVERRIDE),
+    "Skip build tests on windows when running in CI unless overridden",
+)
+@parameterized_class(
```
Comment (jfuss) [inline tests/integration/buildcmd/test_build_cmd.py:1130]: nice!

File: tests/integration/buildcmd/test_build_cmd.py [hunk 3]
```diff
@@ -1122,6 +1123,65 @@ def test_go_must_fail_with_unknown_architecture(self, runtime, code_uri, archite
...
+        ),
+        (
+            "template_build_method_rust_binary.yaml",
+            "Rust/multi-binaries",
+            "function_b",
+            {"req_id": "99", "msg": "Hello FunctionB"},
+        ),
+    ],
+)
+class TestBuildCommand_Rust(BuildIntegRustBase):
```
Comment (jfuss) [inline tests/integration/buildcmd/test_build_cmd.py:1153]: Is Rust buildable in a container? If so, should we have a test for that too?

**Review by jfuss — CHANGES_REQUESTED** [review]

File: appveyor-ubuntu.yml (hunk 1 above)
Comment (hawflau) [reply appveyor-ubuntu.yml:26]: That's a good idea. We might still want to keep cargo tests running in the Canary though (i.e. appveyor-ubuntu.yml and appveyor-windows.yml). But moving those running per PR will prevent each PR from taking longer for the checks.

File: tests/integration/buildcmd/test_build_cmd.py (hunk 3 above)
Comment (hawflau) [reply tests/integration/buildcmd/test_build_cmd.py:1153]: It's buildable in a container but customers will need to provide their own build image.
For now, our `provided.al2` build image doesn't have Rust toolchain installed. We have a plan to provide individual `provided.al2` build images with specific toolchain for each of Go, Rust, but it's not proritized yet.

File: appveyor-ubuntu.yml (hunk 1 above)
Comment (hawflau) [reply appveyor-ubuntu.yml:26]: Made a change to use GH Actions to run Rust integration tests but there are some caveats:
- The GH Actions now can only run on `ubuntu-latest` but not Windows. Reason is I couldn't find a way to switch Windows Docker to use Linux Containers (unlike AppVeyor).
- Therefore, we still need AppVeyor to run these tests in Windows, which is defined in `appvyeor-windows-build-rust.yml`
- We still run the Rust integration tests in Canary (defined in appveyor-ubuntu.yml and appveyor-windows.yml) to keep the completeness of tests in Canary
- The result is - we only remove running Rust tests in the Appveyor make pr check. But the benefit is significant - we avoid adding an extra 22minutes of time to the Appveyor make pr check.
- Downside - it is a bit confusing as tests are spread across in different CI scripts. I'm a bit worried about the possibility of us missing some tests by having none of the CI scripts to pick up these tests.

Tagging @moelasmar as well as he is working on SAM CLI release improvement. For this PR, I'm in favor of running the Rust integration tests in GH Actions to not add extra time in AppVeyor make pr check. But let's keep the downside in mind and be sure to address it when we revisit this topic in SAM CLI release improvement

Comment (jfuss) [reply appveyor-ubuntu.yml:26]: > it is a bit confusing as tests are spread across in different CI scripts. I'm a bit worried about the possibility of us missing some tests by having none of the CI scripts to pick up these tests

Could we make it a single command (within `Makefile`) to make this easier?

**Comment by jfuss** [comment]
@hawflau Once this is merged, let's make sure `continuous-integration/appveyor/pr/windows-build-rust` job get's marked as Required and same for any of the Github Actions for testing.

**Review by moelasmar — APPROVED** [review]
It looks good to me. Just a question shall we add package and deployment integration test cases as per the last action item we decide based on the last issue we faced for Go ?

File: appveyor-windows-build-rust.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
+version: 1.0.{build}
```
Comment (mndeveci) [inline appveyor-windows-build-rust.yml:1]: I think we need this since GH actions are not supporting some stuff for windows?

Comment (hawflau) [reply appveyor-windows-build-rust.yml:1]: yes, unfortunately GH Actions' Windows runner doesn't have the Docker support (switch to Linux containers) that we need. At least I couldn't figure out how to do that as of now.

**Review by mndeveci — APPROVED** [review]
